dep0/lib/03_typecheck/src/alpha_equivalence.cpp

```diff
@@ -0,0 +1,156 @@
+
+struct alpha_equivalence_visitor
+{
+    template <typename T, typename U>
+    requires (not std::is_same_v<T, U>)
+    dep0::expected<std::true_type> operator()(T const& x, U const& y) const
+    {
+        std::ostringstream err;
+        pretty_print<properties_t>(err << '`', x) << '`';
+        pretty_print<properties_t>(err << "is not alpha-equivalent to `", y) << '`';
+        return dep0::error_t(err.str());
+    }
+
+    dep0::expected<std::true_type> operator()(type_t::bool_t const&, type_t::bool_t const&) const { return {}; }
+    dep0::expected<std::true_type> operator()(type_t::unit_t const&, type_t::unit_t const&) const { return {}; }
+    dep0::expected<std::true_type> operator()(type_t::i8_t const&, type_t::i8_t const&) const { return {}; }
+    dep0::expected<std::true_type> operator()(type_t::i16_t const&, type_t::i16_t const&) const { return {}; }
+    dep0::expected<std::true_type> operator()(type_t::i32_t const&, type_t::i32_t const&) const { return {}; }
+    dep0::expected<std::true_type> operator()(type_t::i64_t const&, type_t::i64_t const&) const { return {}; }
+    dep0::expected<std::true_type> operator()(type_t::u8_t const&, type_t::u8_t const&) const { return {}; }
+    dep0::expected<std::true_type> operator()(type_t::u16_t const&, type_t::u16_t const&) const { return {}; }
+    dep0::expected<std::true_type> operator()(type_t::u32_t const&, type_t::u32_t const&) const { return {}; }
+    dep0::expected<std::true_type> operator()(type_t::u64_t const&, type_t::u64_t const&) const { return {}; }
+
+    dep0::expected<std::true_type> operator()(type_t::var_t const& x, type_t::var_t const& y) const
+    {
+        if (x == y)
+            return std::true_type{};
+        else
+        {
+            std::ostringstream err;
+            pretty_print<properties_t>(err << '`', x) << '`';
+            pretty_print<properties_t>(err << " is not alpha-equivalent to `", y) << '`';
+            return dep0::error_t(err.str());
+        }
+    }
+
+    dep0::expected<std::true_type> operator()(type_t::arr_t& x, type_t::arr_t& y) const
+    {
+        if (x.arg_kinds.size() != y.arg_kinds.size())
+        {
+            std::ostringstream err;
+            err << "a function with " << x.arg_kinds.size() << " arguments is not alpha-equivalent to ";
+            err << "a function with " << y.arg_kinds.size();
+            return dep0::error_t(err.str());
+        }
+        auto const not_alpha_equivalent = [&] (std::size_t const i)
+        {
+            auto const print_ordinal = [] (std::ostream& os, std::size_t const i) -> std::ostream&
+            {
+                return os << i << ordinal(i);
+            };
+            std::ostringstream err;
+            pretty_print<properties_t>(print_ordinal(err, i+1) << " argument of type `", x.arg_kinds[i]) << '`';
+            pretty_print<properties_t>(err << " is not alpha-equivalent to argument of type `", y.arg_kinds[i]) << '`';
```

Comment:
once this is using arg_t instead of arg_kind_t, this can read as follows
```suggestion
            pretty_print<properties_t>(print_ordinal(err, i+1) << " argument `", x.arg[i]) << '`';
            pretty_print<properties_t>(err << " is not alpha-equivalent to argument `", y.arg[i]) << '`';
```